### DISCUSSION TITLE
solid-js-router: avoid onMount on each navigation

### DISCUSSION THREAD

**Discussion by tkroll** [solidjs/solid#1054] [Q&A]
Hello!

Just noticed that `onMount` is getting called again on each navigation.
Is there anyway to suppress this? I need `onMount` only the first time the page is visited.

I'm using `hashIntegration` with `NavLink`.

Thanks

**Comment by gregrmunday** [comment]
Hi @tkroll, did you ever find a solution for this?

**Reply by tkroll** [reply]
Hello. I'm sorry but I really can't remember. I usually would update if I found a solution, so I likely worked around it.
Are you still seeing the onMount behavior per navigation?

**Comment by snnsnn** [comment]
Router disposes a component when you navigate away from it. That is reasonable because so many parameters may have been been changed between two visits, otherwise using the previously rendered component would give you static pages.

If you don't want to re-render a component you can wrap it into a detached root using `createRoot` and dispose it manually when you done with it.

A memoized component would also work but there is no way to trigger `onDispose` function on a memoized value.

```typescript
import { Link, Route, Router, Routes, hashIntegration } from '@solidjs/router';
import { Component, createRoot, onMount, JSXElement } from 'solid-js';
import { render } from 'solid-js/web';

const Home = () => {
  onMount(() => console.log('Home Mounted'));
  return <div>Home</div>;
};

const Blog = () => {
  onMount(() => console.log('Blog Mounted'));
  return <div>Blog</div>;
};

const Contact = () => {
  onMount(() => console.log('Contact Mounted'));
  return <div>Contact</div>;
};

let dispose: () => void;
const home = createRoot((disposer) => {
  dispose = disposer;
  return <Home />;
}) as unknown as JSXElement;

const App = () => {
  return (
    <Router source={hashIntegration()}>
      <ul>
        <li><Link href="/home">Home</Link></li>
        <li><Link href="/blog">Blog</Link></li>
        <li><Link href="/contact">Contact</Link></li>
      </ul>
      <Routes>
        <Route path="/home" component={home} />
        <Route path="/blog" component={Blog} />
        <Route path="/contact" component={Contact} />
      </Routes>
    </Router>
  );
};

render(App, document.body);
```

Here is the related SO question: https://stackoverflow.com/a/77295254/7134134
